File: FishWebServer/src/main/kotlin/com/marineseo/fishtank/fishwebserver/model/FishPacket.kt
package com.marineseo.fishtank.fishwebserver.model

import java.nio.ByteBuffer
import java.nio.ByteOrder
import kotlin.experimental.and
import kotlin.experimental.or

const val OP_GET_TEMPERATURE: Short = 1000
const val OP_PIN_IO: Short = 1001
const val OP_READ_DIGIT_PIN: Short = 1002
const val OP_INPUT_ANALOG_PIN: Short = 1003
const val OP_READ_ANALOG_PIN: Short = 1004

const val MAGIC: Short = 31256
const val PACKET_SIZE = 20

private var autoIncId: Int = 0

const val STX: Byte = 0x02
const val ETX: Byte = 0x03
const val DLE: Byte = 0x10

data class FishPacket(
    val id: Int = ++autoIncId,
    val clientId: Int = 0,
    val opCode: Short = 0,
    val pin: Short = 0,
    val pinMode: Short = 0,
    val data: Float = 0f,
    var crc: Short = 0
)

fun FishPacket.makeCrc(): Short {
    return ((id shl 12).toShort() and 0xF000.toShort()) or
            ((clientId shl 8).toShort() and 0x0F00) or
            ((pin.toInt() shl 4).toShort() and 0x00F0) or
            (pinMode and 0x000F)
}

fun FishPacket.isValidate(): Boolean {
    return crc == makeCrc()
}

fun FishPacket.toRawPacket(): ByteArray {
    crc = makeCrc()

    val buffer = ByteBuffer.allocate(PACKET_SIZE * 2)
    buffer.order(ByteOrder.LITTLE_ENDIAN)

    buffer.put(STX)
    write(id, buffer)
    write(clientId, buffer)
    write(opCode, buffer)
    write(pin, buffer)
    write(pinMode, buffer)
    write(data, buffer)
    buffer.put(ETX)
    buffer.putShort(crc)

    buffer.limit(buffer.position())
    buffer.flip()

    return ByteBuffer.allocate(buffer.limit()).apply { put(buffer) }.array()
}

fun ByteArray.toPacket(): FishPacket {
    if(this.size < PACKET_SIZE) throw IllegalArgumentException("Size error. Size must be $PACKET_SIZE at least.")
    if(this[0] != STX) throw IllegalArgumentException("First byte should be $STX but ${this[0]}")
    if(this[size-3] != ETX) throw IllegalArgumentException("Last byte should be $ETX but ${this[size-3]}")

    val crc = ByteBuffer.allocate(2).also {
        it.order(ByteOrder.LITTLE_ENDIAN)
        it.put(this[size-2])
        it.put(this[size-1])
        it.position(0)
    }.short

    // Fetch content. STX~ETX
    val contents = this.filterIndexed { i: Int, _: Byte -> i in 1 until size-3 }.toByteArray()

    val buffer = ByteBuffer.allocate(this.size)
    buffer.order(ByteOrder.LITTLE_ENDIAN)

    var escaped = false
    for(b in contents) {
        if(!escaped && b == DLE){
            escaped = true
            continue
        }
        escaped = false
        buffer.put(b)
    }

    buffer.limit(buffer.position())
    buffer.flip()

    return FishPacket(
        id = buffer.int,
        clientId = buffer.int,
        opCode = buffer.short,
        pin = buffer.short,
        pinMode = buffer.short,
        data = buffer.float,
        crc = crc
    )
}


fun write(value: Number, buffer: ByteBuffer) {
    val size = when(value) {
        is Long -> Long.SIZE_BYTES
        is Int -> Int.SIZE_BYTES
        is Short -> Short.SIZE_BYTES
        is Double -> Double.SIZE_BYTES
        is Float -> Float.SIZE_BYTES
        is Byte -> Byte.SIZE_BYTES
        else -> 0
    }

    val number = if(value is Float) value.toFloat().toRawBits() else value.toInt()
    repeat(size) { index ->
        val b = number.ushr(index * 8).toByte()
        when(b) {
            STX, ETX, DLE -> buffer.put(DLE)
        }

        buffer.put(b)
    }
}

File: Arduino/FishTankNative/FishPacket.cpp
//
// Created by MSSeo on 2022-08-31.
//
#include "FishPacket.h"



/**
 * Write value to target in little endian order.
 * Assume value is variable.
 *
 * @param sizeOfValue
 * @param value value to write to target.
 * @param target target address in buffer.
 * @return count of byte written in this function.
 */
int FishPacket::write(const unsigned char* value, int sizeOfValue, unsigned char* target) {
    int idx = 0;

    for(int i=0; i<sizeOfValue; i++) {
        // Meta data should be escaped.
        switch (value[i]) {
            case STX:
            case ETX:
            case DLE:
                // Add DLE for escape.
                target[idx++] = DLE;
                break;
        }

        target[idx++] = value[i];
    }

    return idx;
}

/**
 * Serialize this packet to byte array.
 *
 * @param bff buffer to serialize.
 * @return byte size of serialized packet.
 */
int FishPacket::serializePacket(unsigned char bff[]) {
    // Little-endian
    int index = 0;
    bff[index++] = STX;

    crc = makeCrc();

    index += write((unsigned char *)&id, 4, bff+index);
    index += write((unsigned char *)&clientId, 4, bff+index);
    index += write((unsigned char *)&opCode, 2, bff+index);
    index += write((unsigned char *)&pin, 2, bff+index);
    index += write((unsigned char *)&pinMode, 2, bff+index);
    index += write((unsigned char *)&data, 4, bff+index);

    bff[index++] = ETX;
    unsigned char* cp = (unsigned char *)&crc;
    bff[index++] = cp[0];
    bff[index++] = cp[1];

    return index;
}

/**
 *
 * @param value
 * @param sizeOfValue
 * @param buffer
 * @return byte size of read from buffer.
 */
int FishPacket::read(unsigned char* value, int sizeOfValue, const unsigned char* buffer) {
    int idx = 0;
    int vIdx = 0;

    bool escaped = false;
    while(true) {
        unsigned char b = buffer[idx++];
        if(!escaped && b == DLE) {
            escaped = true;
            continue;
        }

        escaped = false;
        value[vIdx++] = b;
        if(vIdx >= sizeOfValue) break;
    }
    return idx;
}

int FishPacket::deSerializePacket(unsigned char bff[]) {
    // Little-endian
    int index = 0;
    if(bff[index++] != STX) return -1;

    index += read((unsigned char *)&id, 4, bff+index);
    index += read((unsigned char *)&clientId, 4, bff+index);
    index += read((unsigned char *)&opCode, 2, bff+index);
    index += read((unsigned char *)&pin, 2, bff+index);
    index += read((unsigned char *)&pinMode, 2, bff+index);
    index += read((unsigned char *)&data, 4, bff+index);
    index += read((unsigned char *)&crc, 2, bff+index);

    if(bff[index++] != ETX) return -1;

    return index;
}

void FishPacket::clear() {
    id = 0;
    clientId = 0;
    opCode = 0;
    pin = 0;
    pinMode = 0;
    data = 0;
    crc = 0;
}

bool FishPacket::validateCrc() {
    return makeCrc() == crc;
}

unsigned int FishPacket::makeCrc() {
    int localCrc = (id << 12) & 0xF000;
    localCrc |= (clientId << 8) & 0x0F00;
    localCrc |= (pin << 4) & 0x00F0;
    localCrc |= pinMode & 0x000F;
    return localCrc;
}

